lib/auth/autoupdate/autoupdatev1/service.go

```diff
@@ -589,3 +608,78 @@ func (s *Service) emitEvent(ctx context.Context, e apievents.AuditEvent) {
 		)
 	}
 }
+
+// Those values are arbitrary, we will want to increase them as we test. We will also want to modulate them based on the
+// cluster context. We don't want people to craft schedules that can't realistically finish within a week on Cloud as
+// we usually do weekly updates. However, self-hosted users can craft more complex schedules, slower rollouts, and shoot
+// themselves in the foot if they want.
+const (
+	maxGroupsTimeBasedStrategy        = 20
+	maxGroupsHaltOnErrorStrategy      = 10
+	maxGroupsHaltOnErrorStrategyCloud = 4
+	maxWaitDaysCLoud                  = 3
+)
+
+var (
+	cloudGroupUpdateDays = []string{"Mon", "Tue", "Wed", "Thu"}
+)
+
+// validateServerSideAgentConfig validates that the autoupdate_config.agent spec meets the cluster rules.
+// Rules may vary based on the cluster, and over time.
+//
+// This function should not be confused with api/types/autoupdate.ValidateAutoUpdateConfig which validates the integrity
+// of the resource and does not enforce potentially changing rules.
+func (s *Service) validateServerSideAgentConfig(config *autoupdate.AutoUpdateConfig) error {
+	// We must check resource integrity before, because it makes no sense to try to enforce rules on an invalid resource.
+	// The generic backend service will likely check integrity again, but it's not a large performance problem.
+	err := update.ValidateAutoUpdateConfig(config)
+	if err != nil {
+		return trace.Wrap(err, "validating autoupdate config")
+	}
+
+	var maxGroups int
+	isCloud := modules.GetModules().Features().Cloud
```

Comment:
We may want to relax these restrictions for some Cloud customers on a case-by-case basis. Would it be tricky to add an "unrestricted" feature? (Could be a separate PR)

Reply:
We will likely want to rely on entitlements. I'm not familiar with this mechanism and will have to ask cloud devs how to do this. If it's not easy, I'll just glue an ugly environment variable.